lib/Crypto/doc/sailfish-crypto-overview.qdoc

```diff
@@ -0,0 +1,218 @@
+of usage of the Sailfish OS Crypto Library as well as the Sailfish OS Secrets
+Library.  Please see those for complete, working examples.
+
+Some snippets showing commonly-required functionality are included below.
+
+\section4 Generating a securely-stored symmetric cipher key
+
+This snippet shows how to generate a symmetric cipher key which will be stored
+securely by the Sailfish OS Secrets and Crypto Framework. It assumes that a
+secure storage collection has previously been created, in which to store the
+key. Please see the Sailfish OS Secrets Library documentation for information
+about how to create such a collection.
+
+First, the client defines the algorithm, supported block modes, supported
+padding schemes, and supported operations in a template key.  That template key
+is given an identifier which includes the name of the key and the name of the
+secure storage collection in which it should be stored, and is then passed as a
+parameter when invoking the \tt{generateStoredKey()} method, which results in an
+IPC call to the secure Sailfish OS Secrets and Crypto Framework system service.
+
+The Sailfish OS Secrets and Crypto Framework system service will then delegate
+the operation to the specified crypto plugin, which in turn will generate a
+full symmetric key based upon the given template, and store it securely.  A
+reference key (that is, a key containing a valid identifier and metadata, but
+no secret key data) will be returned to the client application.
+
+Note that these operations are all asynchronous, however in the snippet we
+force the operation to be synchronous by calling the \tt{waitForFinished()}
+method on the reply object.  In practice, the client application should instead
+use a watcher object which will notify when the operation is complete.
+
+\code
+// Set the key template metadata.
+Sailfish::Crypto::Key keyTemplate, symmetricKeyReference;
+keyTemplate.setAlgorithm(Sailfish::Crypto::Key::Aes256);
+keyTemplate.setOrigin(Sailfish::Crypto::Key::OriginDevice);
+keyTemplate.setBlockModes(Sailfish::Crypto::Key::BlockModeCBC);
+keyTemplate.setEncryptionPaddings(Sailfish::Crypto::Key::EncryptionPaddingNone);
+keyTemplate.setSignaturePaddings(Sailfish::Crypto::Key::SignaturePaddingNone);
+keyTemplate.setDigests(Sailfish::Crypto::Key::DigestSha256);
+keyTemplate.setOperations(Sailfish::Crypto::Key::Encrypt
+                         |Sailfish::Crypto::Key::Decrypt);
+
+// Set the identifier for the key.
+// This assumes the existence of an "ExampleCollection" secure storage
+// collection, in which the key will be stored.
+// See Sailfish::Secrets::SecretManager::createCollection().
+keyTemplate.setIdentifier(QStringLiteral("ExampleKey"),
+                          QStringLiteral("ExampleCollection"));
+
+// Ask the system service to generate and store the key securely.
+Sailfish::Crypto::CryptoManager cm;
+QDBusPendingReply<Sailfish::Crypto::Result, Sailfish::Crypto::Key>
```

Comment:
If you look It is rare for DBus parts to leak outside the client-side Qt C++ wrapper API, usually the API provides all the necessary structures and signals to track the status of the requests. Might be a non-issue, just wondering.

Reply:
I'd never leak outside the transport. I'd even hide everything so that the user would never know that the data is even passed somewhere. The user must not know that. My opinion. Also, what would the user have if we suddenly decide to change the DBus to anything else? Or just to another dbus library? All the user code will need to be changed. Do we really want it? Also, as I told before, it makes the user code more complicated since the user must everytime to unpack the result. Useless work. But it will be everywhere in the user code. Probably, it is so useless work that everyone will write his own abstractions (wrappers) around that work if it is possible. I'd do that. I'd do that even due to another reasons such as abstractions of future changes of the API (eg we are moving from dbus to anything else which need not to be unpacked). So, I'd simply not leak outside anything which is not related to the purpose of the framework.

No offense please, I hope I was polite enough. Want to help. <3

Reply:
No offense taken, and you're right.  However, given that documentation became a high priority item, I thought it best to write docs for what currently exists immediately, rather than spending potentially large amount of time designing and implementing a different client API - because (and here's the important thing) I believe that the MOST IMPORTANT thing we have to resolve right now is: does the API support the required features, and are the parameters to the operations correct.  Documenting the current API helps us to resolve these questions (I hope...).

Next week I will take your suggestion and start prototyping a client API (which internally wraps around the current Secret/CryptoManager APIs which will become private API) which hides the transport and also uses some form of per-parameter accessor/mutator request data object.

Reply:
Good! Okay.:)